Server/ConcordiaCurriculumManager/Controllers/GroupController.cs

```diff
@@ -79,16 +105,111 @@ public async Task<IActionResult> AddUserToGroup(Guid groupId, Guid userId)
+        }
+
         var result = await _groupService.RemoveUserFromGroup(userId, groupId);
         if (result)
         {
             return Ok();
         }
         return NotFound();
     }
+
+    [HttpPost("{groupId}/masters/{userId}")]
+    [Authorize]
+    [SwaggerResponse(StatusCodes.Status200OK, "User added to group masters successfully")]
+    [SwaggerResponse(StatusCodes.Status404NotFound, "User is not a member of this group or is already a group master")]
+    public async Task<IActionResult> AddGroupMaster(Guid groupId, Guid userId)
+    {
+        Guid currentUserId;
+        try
+        {
+            currentUserId = Guid.Parse(_userService.GetCurrentUserClaim(Claims.Id));
+        }
+        catch (Exception e)
+        {
+            var message = "Unexpected error occured while retrieving user information.";
+            _logger.LogWarning($"${message}: {e.Message}");
+            return Problem(
+                title: message,
+                detail: e.Message,
+                statusCode: StatusCodes.Status500InternalServerError);
+        }
+
+        if (!await IsAdminOrGroupMaster(currentUserId, groupId))
+        {
+            return StatusCode(StatusCodes.Status403Forbidden, "You must be an admin or a group master to make changes to this group");
+        }
+
+        var result = await _groupService.AddGroupMaster(userId, groupId);
+        if (result)
+        {
+            return Ok();
+        }
+        return NotFound();
+    }
+
+    [HttpDelete("{groupId}/masters/{userId}")]
+    [Authorize]
+    [SwaggerResponse(StatusCodes.Status200OK, "User removed from group masters successfully")]
+    [SwaggerResponse(StatusCodes.Status404NotFound, "User is not a member of this group or is not a group master")]
+    public async Task<IActionResult> RemoveGroupMaster(Guid groupId, Guid userId)
+    {
+        Guid currentUserId;
+        try
+        {
+            currentUserId = Guid.Parse(_userService.GetCurrentUserClaim(Claims.Id));
+        }
+        catch (Exception e)
+        {
+            var message = "Unexpected error occured while retrieving user information.";
+            _logger.LogWarning($"${message}: {e.Message}");
+            return Problem(
+                title: message,
+                detail: e.Message,
+                statusCode: StatusCodes.Status500InternalServerError);
+        }
+
+        if (!await IsAdminOrGroupMaster(currentUserId, groupId))
+        {
+            return StatusCode(StatusCodes.Status403Forbidden, "You must be an admin or a group master to make changes to this group");
+        }
+
+        var result = await _groupService.RemoveGroupMaster(userId, groupId);
+        if (result)
+        {
+            return Ok();
+        }
+        return NotFound();
+    }
+
+    public async Task<bool> IsAdminOrGroupMaster(Guid currentUserId, Guid groupId)
```

Comment:
Btw, the recommended way to perform authorization, for such logic, is using [policy-based authorization](https://learn.microsoft.com/en-us/aspnet/core/security/authorization/policies?view=aspnetcore-7.0). This is to separate the authorization logic from the actual service/endpoint logic and allow ease of reuse. We can refactor it later so you don't have to worry about it for now.

Reply:
Even worse, it's resource-based authorization:
https://code-maze.com/aspnetcore-resource-based-authorization/